dev/buildtool/spin_commands.py

```diff
@@ -312,39 +291,22 @@ def promote_spin(self, repository):
     git_dir = repository.git_dir
     git = self.__scm.git
 
-    match = re.match(r'(\d+)\.(\d+)\.(\d+)-\d+', self.__gate_version)
-    gate_major = match.group(1)
-    gate_min = match.group(2)
-    if match is None:
-      raise_and_log_error(
-          ConfigError('gate version {version} is not X.Y.Z-<buildnum>'
-                      .format(version=self.__gate_version)))
-
     if len(self.__spinnaker_version.split('.')) != 3:
       raise_and_log_error(
           ConfigError('Expected spinnaker version in the form X.Y.Z-N'))
 
-    # Note: spin CLI is coupled to the Gate major and minor version.
-    # Gate is a routing server, so features and breaking changes in Gate
-    # must be reflected in spin since it is a client. We pin only the major
-    # and minor versions so fixes (thus patch version) are decoupled between
-    # the two.
-    patch = '0' # Patch is reset on a new Gate major or minor.
-    tag_matcher = re.compile(r'version-{maj}.{min}.(\d+)'
-                             .format(maj=gate_major, min=gate_min))
-    tags = git.fetch_tags(git_dir)
-    tag_matches = [tag_matcher.match(t) for t in tags if tag_matcher.match(t)]
-    if tag_matches:
-      patch_versions = [int(m.group(1)) for m in tag_matches]
-      max_patch = max(patch_versions)
-      last_tag = 'version-{maj}.{min}.{max_patch}'.format(maj=gate_major,
-                                                          min=gate_min,
-                                                          max_patch=max_patch)
-      self.__no_changes = git.query_local_repository_commit_id(git_dir) == git.query_commit_at_tag(git_dir, last_tag)
-      patch = str(max_patch + 1)
-
-    self.__stable_version = '{major}.{minor}.{patch}'.format(
-        major=match.group(1), minor=match.group(2), patch=patch)
+    next_spin_semver = bump_spin_patch(git, git_dir, self.__gate_version)
+    gate_major = next_spin_semver.major
+    gate_min = next_spin_semver.minor
+    last_patch = str(max(next_spin_semver.patch - 1, 0))
+    self.__stable_version = next_spin_semver.to_version()
+    logging.info('calculated new stable version: {}'
+                 .format(self.__stable_version))
+    last_tag = 'version-{maj}.{min}.{patch}'.format(maj=gate_major,
```

Comment:
Im confused here. Why do you use next last maj/minor for the last tag?
If you just rolled one of these, then it is different from the last one.

Reply:
Major and minor are pinned to the gate version's major and minor and will be static.